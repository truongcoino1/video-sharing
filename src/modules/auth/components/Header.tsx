/* eslint-disable @next/next/no-img-element */
"use client"
import React, { memo } from "react";
import Link from "next/link";
import Login from "./Login";
import UserAction from "./UserAction";
import { useAuthContext } from "../context";
import LoginPlaceholder from './LoginPlaceholder'

const PATH_HOME = "/";

const Header = () => {
  const { currentUser, loading } = useAuthContext();

  return (
    <header className="header-container p-[16px] lg:p-0 lg:h-[66px] lg:flex lg:justify-between lg:items-center border-b border-gray-300">
      <h1 className="title text-[36px] font-semibold mb-[8px] lg:mb-0">
        <Link className="no-style hover:text-black text-light-high flex" href={PATH_HOME}>
          <img src="/funny-movie.png" className="w-[48px] h-[48px] mr-12" alt="" />
          Funny Movies
        </Link>
      </h1>
      {loading ? <LoginPlaceholder/> : currentUser ? <UserAction email={currentUser.email} /> : <Login />}
    </header>
  );
};

export default memo(Header);
