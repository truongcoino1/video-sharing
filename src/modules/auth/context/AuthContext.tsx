"use client";

import { Loading, setToken } from "@/modules/base";
import React, { useEffect, useState, useContext } from "react";
import { User } from "../types/auth";
import { useRouter } from "next/navigation";

type AuthContextProps = {
  currentUser: User | null;
  setCurrentUser: (user: User | null) => void;
  setLoading: (loading: boolean) => void;
  loading: boolean;
};
export const AuthContext = React.createContext<AuthContextProps>({
  currentUser: null,
  setCurrentUser: (_: User | null) => {},
  setLoading: (_: boolean) => {},
  loading: false
});

export const useAuthContext = () => useContext<AuthContextProps>(AuthContext);

interface AuthProviderProps {
  children: React.ReactNode;
}

export const AuthProvider = ({ children }: AuthProviderProps) => {
  const [currentUser, setCurrentUser] = useState<User | null>(null);
  const [loading, setLoading] = useState(true);
  const router = useRouter();

  useEffect(() => {
    const authString = localStorage.getItem("auth");
    if (authString) {
      const auth = JSON.parse(authString);
      setCurrentUser(auth);
      setToken(auth.token);
    } else {
      router.push("/");
    }
    setLoading(false);
    console.log('vao day roi ne')
  }, []);

  return (
    <AuthContext.Provider value={{ loading, currentUser, setLoading, setCurrentUser }}>
      {children}
    </AuthContext.Provider>
  );
};
