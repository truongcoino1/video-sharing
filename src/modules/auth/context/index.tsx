"use client";

import { ReactNode } from 'react'
import { AuthProvider} from './AuthContext'

const AllProviders = ({ children }: { children: ReactNode }) => {
  return <AuthProvider>{children}</AuthProvider>
}

export {AllProviders}
export * from './AuthContext'