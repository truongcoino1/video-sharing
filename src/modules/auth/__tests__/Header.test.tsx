import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { AuthContext } from "../context/AuthContext";
import Header from "../components/Header";

jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: jest.fn(),
  }),
  usePathname: () => "/",
}));
 //@ts-ignore
 global.setImmediate = jest.useRealTimers;
const setUpNotLogin = () => {
  return render(
    <AuthContext.Provider
      value={{
        currentUser: null,
        setLoading: jest.fn(),
        setCurrentUser: jest.fn(),
        loading: false,
      }}
    >
      <Header />
    </AuthContext.Provider>
  );
};

const setUpLogin = () => {
  return render(
    <AuthContext.Provider
      value={{
        currentUser: {
          email: "truongvk",
          token: "123",
          id: 1,
          password: "123",
          refreshToken: "123",
        },
        setLoading: jest.fn(),
        setCurrentUser: jest.fn(),
        loading: false,
      }}
    >
      <Header />
    </AuthContext.Provider>
  );
};

describe("Header", () => {
  afterEach(cleanup);
 
  it("Header render without crash", async () => {
    const result = setUpNotLogin();
    const elm = result.container.querySelector(".header-container");
    expect(elm).toBeInTheDocument();
  });

  it("Header show correct title", () => {
    const result = setUpNotLogin();
    const elm = result.container.querySelector(".title");
    expect(elm?.textContent).toBe("Funny Movies");
  });

  it("Header show login", () => {
    const result = setUpNotLogin();
    const elm = result.container.querySelector(".login-container");
    expect(elm).toBeInTheDocument();
  });

  it("Header show userAction", () => {
    const result = setUpLogin();
    const elm = result.container.querySelector(".user-action-container");
    expect(elm).toBeInTheDocument();
  });

  it("Header show userAction after login/register success", () => {
    const result = setUpNotLogin();

    // Firstly, display loign
    const container = result.container;
    const loginElm = result.container.querySelector(".login-container");
    expect(loginElm).toBeInTheDocument();

    const inputEmail = container.querySelector(
      'input[type="email"]'
    ) as Element;
    const inputPassword = container.querySelector(
      'input[type="password"]'
    ) as Element;
    const buttonSubmit = container.querySelector(
      'button[type="submit"]'
    ) as Element;

    fireEvent.change(inputEmail, {
      target: { value: "test_email1@gmail.com" },
    });
    fireEvent.change(inputPassword, { target: { value: "abc123451" } });
    expect(buttonSubmit).not.toBeDisabled();

    fireEvent.submit(buttonSubmit);

    // After login success
    const actionElm = result.container.querySelector(".user-action-container");
    waitFor(() => expect(actionElm).toBeInTheDocument());
  });

  it("should render correct Header", () => {
    const header = render(<Header />);
    expect(header).toMatchSnapshot();
  });
});
