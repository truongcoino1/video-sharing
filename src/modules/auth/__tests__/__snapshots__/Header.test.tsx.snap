// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correct Header 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="header-container lg:h-[66px] lg:flex lg:justify-between lg:items-center border-b border-gray-300"
      >
        <h1
          class="title text-[36px] font-semibold"
        >
          <a
            class="no-style hover:text-black text-light-high flex"
            href="/"
          >
            <img
              alt=""
              class="w-[48px] h-[48px] mr-12"
              src="/funny-movie.png"
            />
            Funny Movies
          </a>
        </h1>
        <div
          class="login-container"
        >
          <form
            class="login-form flex-col flex lg:flex-row"
          >
            <input
              aria-label="email"
              class="lg:mr-8 h-[40px] px-8 mt-12 lg:mt-0 border rounded-medium border-gray-300"
              name="email"
              placeholder="Email"
              type="email"
              value=""
            />
            <input
              aria-label="password"
              class="lg:mr-8 h-[40px] px-8 border mt-12 lg:mt-0 rounded-medium border-gray-300"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
            <button
              class="px-8 lg:mt-0 h-[40px] mt-12 rounded-medium flex items-center justify-center opacity-50 cursor-not-allowed bg-dark-200 text-dark-high "
              disabled=""
              type="submit"
            >
              Login / Register
            </button>
          </form>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="header-container lg:h-[66px] lg:flex lg:justify-between lg:items-center border-b border-gray-300"
    >
      <h1
        class="title text-[36px] font-semibold"
      >
        <a
          class="no-style hover:text-black text-light-high flex"
          href="/"
        >
          <img
            alt=""
            class="w-[48px] h-[48px] mr-12"
            src="/funny-movie.png"
          />
          Funny Movies
        </a>
      </h1>
      <div
        class="login-container"
      >
        <form
          class="login-form flex-col flex lg:flex-row"
        >
          <input
            aria-label="email"
            class="lg:mr-8 h-[40px] px-8 mt-12 lg:mt-0 border rounded-medium border-gray-300"
            name="email"
            placeholder="Email"
            type="email"
            value=""
          />
          <input
            aria-label="password"
            class="lg:mr-8 h-[40px] px-8 border mt-12 lg:mt-0 rounded-medium border-gray-300"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <button
            class="px-8 lg:mt-0 h-[40px] mt-12 rounded-medium flex items-center justify-center opacity-50 cursor-not-allowed bg-dark-200 text-dark-high "
            disabled=""
            type="submit"
          >
            Login / Register
          </button>
        </form>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
